spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,298 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.sources.v2.DataSourceOptions;
+import org.apache.spark.sql.sources.v2.reader.streaming.MicroBatchReader;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A mirco-batch based Spark Structured Streaming reader for Iceberg table. It will track the added
+ * files and generate tasks per batch to process newly added files. By default it will process
+ * all the newly added files to the current snapshot in each batch, user could also set this
+ * configuration "max-files-per-trigger" to control the number of files processed per batch.
+ */
+class StreamingReader extends Reader implements MicroBatchReader {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReader.class);
+  private static final int DEFAULT_MAX_FILES_PER_TRIGGER = 1000;
+
+  private static final Comparator<FileScanTask> COMPARATOR = (left, right) -> {
+    int pathCompare = left.file().path().toString().compareTo(right.file().path().toString());
+    int startCompare = Long.compare(left.start(), right.start());
+    int lengthCompare = Long.compare(left.length(), right.length());
+    return pathCompare == 0 ? (startCompare == 0 ? lengthCompare : startCompare) : pathCompare;
+  };
+
+  private StreamingOffset startOffset;
+  private StreamingOffset endOffset;
+
+  private final Table table;
+  private final int maxFilesPerTrigger;
+  private final long splitSize;
+  private final int lookback;
+  private final long openFileCost;
+  private final Long startSnapshotId;
+
+  /**
+   * Utility class to track the snapshotId, task and index of task within this snapshot.
+   */
+  @VisibleForTesting
+  static class IndexedTask {
+    private final long snapshotId;
+    private final FileScanTask task;
+    private final int index;
+
+    IndexedTask(long snapshotId, FileScanTask task, int index) {
+      this.snapshotId = snapshotId;
+      this.task = task;
+      this.index = index;
+    }
+
+    long snapshotId() {
+      return snapshotId;
+    }
+
+    FileScanTask task() {
+      return task;
+    }
+
+    int index() {
+      return index;
+    }
+  }
+
+  StreamingReader(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryptionManager,
+                  boolean caseSensitive, DataSourceOptions options) {
+    super(table, io, encryptionManager, caseSensitive, options);
+
+    this.table = table;
+    this.maxFilesPerTrigger =
```

Comment:
Should this use a maximum number of files or a maximum number of tasks? It seems like limiting tasks would result in a more stead workload. By limiting files, this will process less when files are small and more when files are large. By limiting tasks after combining and splitting, the total amount of data processed is more uniform.

Reply:
> By limiting tasks after combining and splitting, the total amount of data processed is more uniform.
I think the difficulty to use tasks rather than files to do rate limit is that: 

1. the semantics of rate limit is related to the definition of offset. By changing to use tasks, how do we define the offset, especially when calling `planTasks` on all the files among snapshots. I think we can only call `planTasks` on each snapshot, and use snapshot index to track tasks per snapshot, this is what I did in the first version of this PR, the problem you mentioned above still exists when data is not balanced among snapshots.
2. I'm not so sure if the algorithm of `binPacking` is stable, if it will be changed due to the file number and size, then it will potentially cause the inconsistency.

Reply:
To answer your questions:

1. We could define the offset in terms of tasks, not files. That works as long as the packing is stable.
2. I see no reason why the algorithm here would change, but we would want to make sure the settings (target size, lookback) are part of the job state. When a job resumes and recalculates splits, we would want the same packing settings.

I could go either way on this. I think usability would be better if it used tasks and not files because it would split/combine and that balances work across batches better. But it is certainly easier to maintain if we use files and don't need to worry about the stability of bin packing settings across runs in addition to file order.

Reply:
What about defining the micro-batch size in terms of data size and planning tasks after that?